### PR TITLE
Parametric "take! and store" message again

### LINKED ISSUE #740: restore parametric covariance msgs
broken during #459 work in #739 

2e977a1 

xref: https://github.com/JuliaRobotics/IncrementalInference.jl/commit/2e977a1e4073d76ab8c94a43fd6f8098782a19e8#r39887387

### PR DESCRIPTION
**Pull request by Affie**
Close #740

### CODE REVIEW

**Comment by Affie** [comment]
Let's discuss this in #855

**Comment by Affie** [comment]
The main difference here is:
- `fetch` vs `take!` #855
- storing channels in `edges` vs `nodes` (easily changed)

**Comment by Affie** [comment]
The point of this pr is to get all the current parametric features working again and then change from there.

**Comment by Affie** [comment]
I think I understand the fundamental difference that still remains between parametric and mm CSMs.
Issues #855 and #675 goes hand in hand. 
- fetch is easier with channels in nodes  
- take! is easier with channels in edges 

I don't think we are going to agree on what strategy to follow, so why not have both for a while. After the down part of #459 is done, it can look close enough with only function names different for fetch vs put!.

**Comment by dehann** [comment]
How far is this work (in the off chance this PR does get merged) away from changing PCSM onto a `fetch` model, so that it matches CSM? https://github.com/JuliaRobotics/IncrementalInference.jl/issues/855#issuecomment-679921053

**Comment by Affie** [comment]
> ... changing PCSM onto a fetch model ...

Maybe I'm missing something, but I don't see that as a good option. To do that I need to add conditions to wait on. Why do that if you can use the channels directly?

File: src/JunctionTree.jl [hunk 1]
```diff
@@ -756,6 +756,16 @@ function initTreeMessageChannels!(tree::MetaBayesTree)
...
+function refTreeMessageChannelsFromBTND!(tree::BayesTree)
+  for inci in tree.bt.inclist
+    for e in inci
+        downMsg = e.target.data.dwnMsgChannel
+        upMsg = e.target.data.upMsgChannel
+        push!(tree.messages, e.index=>(upMsg=upMsg,downMsg=downMsg))
+    end
+  end
+  return nothing
+end
```
Comment (Affie) [inline src/JunctionTree.jl:768]: Based on https://github.com/JuliaRobotics/IncrementalInference.jl/issues/675#issuecomment-679954452
A container that keeps all channels in one place is also handy for debugging and monitoring. we can also send termination messages to cleanly exit as with PCSM

**Comment by dehann** [comment]
reluctantly removing the blocker, with condition that we work to get both PCSM and CSM to use the same model from your examples #675 -- CSM will soon only be model 1. and sounds like PCSM is model 2.

For completeness, CSM used to be somewhere between model 1 and 3, and after #459 will be only model 1.

**Comment by Affie** [comment]
> reluctantly removing the blocker, with condition that we work to get both PCSM and CSM to use the same model from your examples #675 -- CSM will soon only be model 1. and sounds like PCSM is model 2.

Yes,  That is correct PCSM is 2.

Condition is fine, so I'm going to merge this as is.

Just to recap on our conversation:

The difference between the 2 designs is `fetch` vs `take!`
- fetch is likely to require conditions as well (as in current CSM)
- take! will likely not require additional conditions, especially if an n-stroke solve works correctly

**Pull request merged**